pkg/lexer/state.go

```diff
@@ -0,0 +1,365 @@
+					l.insertSemi = t.InsertSemi()
+
+					lexStmt(l)
+					goto insertSemi
+				}
+
+				// commands don't start with a keyword
+				l.emit(cmdOpLookup(word))
+			}
+
+			lexCmd(l)
+
+			// semicolon insertion
+		insertSemi:
+			l.emit(token.SEMICOLON)
+		}
+	}
+}
+
+func isAlphabet(r rune) bool {
+	return r >= 'A' && r <= 'Z' || r >= 'a' && r <= 'z'
+}
+
+func lexStmt(l *lexer) {
+	for {
+		l.consume()
+
+		switch {
+		case l.ch == '\n':
+			if l.insertSemi {
+				// if semicolon is inserted the statement ends
+				return
+			}
+
+			// ignore all space
+			consumeAllSpace(l)
+
+		case unicode.IsSpace(l.ch):
+			if l.insertSemi {
+				// semicolon to be inserted
+				// do not ignore newlines
+				consumeSpace(l)
+				break
+			}
+
+			// no semicolon to be inserted
+			// ignore all space runes
+			consumeAllSpace(l)
+
+		case isIdentStart(l.ch):
+			t := lexIdent(l)
+			l.insertSemi = t.InsertSemi()
+
+		case unicode.IsDigit(l.ch):
+			lexNum(l)
+			// semicolon should be inserted after a number
+			l.insertSemi = true
+
+		case l.ch == '"':
+			lexString(l)
+			// semicolon should be inserted after a string
+			l.insertSemi = true
+
+		// all operator starting runes are themselves operators
+		case token.IsOperator(string(l.ch)):
+			t := lexStmtOp(l)
+			l.insertSemi = t.InsertSemi()
+
+		case l.ch == '#':
+			// line comment
+			lexComment(l)
+
+		case l.ch == eof:
+			return // will be handled by lexBase
+
+		default:
+			// rune not supported inside statements
+			l.emit(token.ILLEGAL)
+		}
+	}
+}
+
+func lexIdent(l *lexer) token.TokenType {
+	for isIdent(l.peek()) {
+		l.consume()
+	}
+
+	// lookup the token type of literal
+	t := token.Lookup(l.literal())
+	l.emit(t)
+	return t
+}
+
+func isIdentStart(r rune) bool {
+	return r == '_' || unicode.IsLetter(r)
+}
+
+func lexNum(l *lexer) {
+	// TODO: support more number types
+	for unicode.IsDigit(l.peek()) {
+		l.consume()
+	}
+
+	l.emit(token.FLOAT)
+}
+
+func (l *lexer) makeOp(target rune, pass token.TokenType, fail token.TokenType) token.TokenType {
+	if l.peek() == target {
+		l.consume()
+		return pass
+	}
+
+	return fail
+}
+
+func lexStmtOp(l *lexer) token.TokenType {
+	var t token.TokenType
+	switch l.ch {
+	case '+':
+		t = l.makeOp('=', token.ADD_ASSIGN, token.ADD)
+	case '-':
+		t = l.makeOp('=', token.SUB_ASSIGN, token.SUB)
+	case '*':
+		t = l.makeOp('=', token.MUL_ASSIGN, token.MUL)
+	case '/':
+		t = l.makeOp('=', token.QUO_ASSIGN, token.QUO)
+	case '%':
+		t = l.makeOp('=', token.REM_ASSIGN, token.REM)
+	case '&':
+		t = l.makeOp('&', token.LAND, token.AND)
+
+		if t == token.LAND {
+			break
+		}
+
+		t = l.makeOp('^', token.AND_NOT, token.AND)
+
+		e := token.AND_NOT_ASSIGN
+		if t == token.AND {
+			e = token.AND_ASSIGN
+		}
+
+		t = l.makeOp('=', e, t)
+	case '|':
+		t = l.makeOp('|', token.LOR, token.OR)
+
+		if t == token.OR {
+			t = l.makeOp('=', token.OR_ASSIGN, token.OR)
+		}
+	case '^':
+		t = l.makeOp('=', token.XOR_ASSIGN, token.XOR)
+	case '<':
```

Comment:
why is this being handled separately?

Reply:
I don't understand what you mean. Could you elaborate?

Reply:
I don't understand what you mean, could you elaborate?

Reply:
There are two places this case is being handled, one that is marked above and [here](https://github.com/raklaptudirm/mash/blob/92934a0171faf526f7bd708f9462b2df809d1d94/pkg/lexer/state.go#L282) which is as a string (why as a string? All you need to check is that whether this is a token (doesn't matter whether its a cmd Operator or not so this function doesn't make sense to me)

Reply:
You see, commands have a different set of operators than statements, so I can't emit them both in the same place. That will either classify non-command operators like `=` as operators, or classify real statement operators as strings.

Reply:
Why are there different set of tokens for these cases? This is exactly what I mean when I say that you're mixing parsers logic into the lexer.

Note two things:
Lexer is only responsible for tokenising the input/source
Parser is only responsible for making sense of the tokens provided by the lexer to convert them into expressions which is done by using the language grammar.

This is a fundamental difference between token and parser. Token doesn't and shouldn't need to know the language grammar other than if a substring is an input then what token does this represent.

What is currently happening is that you're trying to force the language grammar level details into the lexer if you have more than one token representing the same stream of input. Remember, token is grammar agnostic, it doesn't care whether what it is tokenising is the correct thing or not, as long as it can tokenise something, it will and if not, it will tell the parser that this input is not a valid token.

Reply:
Finally, why is there is distinction between a statement is allowed to not have a space but a command isn't.

I believe you'll agree that

```
some_command -b>>README.md
```
is valid, as well as 

```
for var in 1,2,3;do echo $var;done;
```
is also valid, there is no distinction here - the reason being that it doesn't matter whether there is space or not since space has no meaning other than to differentiate between two input that could combine to make one big input.

Reply:
Take this two for for example:
```
let a = b + c
echo b + c
```
In the first, `+` is the addition operator, and should be treated like one. In the second case though `+` is just an argument, a string.

Also note this:
```
let a = b&&c
echo a&&b
```
In the first, `b&&c` may be together, but they are three different tokens, `a`, `&&`, and `c`. In the second case though, they are a single argument, a string.

The lexer needs to tokenize them according to those rules which I just said. Understanding what type a token is and how to tokenize it exactly is indeed in the parser's domain.

Reply:
Ok, I understand where you're coming from, so let me take these as your examples
```
let a = b + c
echo b + c
```

and add the following

```
echo b + c>>README
```
What happens now?

Also, the echo command is not taking in one input in fact, its taking in all inputs as arguments and concatenating them (look at the source for any early shells)

In the first case `let a = b + c`, if we tokenise the normal way, we would have token.LET, token.IDENT, token.EQ, token.IDENT, token.ADD, token.IDENT.

In the second case, `echo b + c` would also be tokenised the same way, token.IDENT, token.IDENT, token.ADD, token.IDENT, what it will do instead is that when the parser looks at the first token.IDENT (it figures out that this is a command) further to this all other operators (such as non command tokens (ADD, SUBTRACT etc) would just become the normal IDENT's as well. This makes it straight forward to make sure that operators like these `>>` which are command level operators don't get ignored by the parser.

Reply:
Does that make sense? Your thinking was correct that we need to have these distinctions, but I think the place to make those distinctions is not the lexer package.

Reply:
> In the second case, `echo b + c` would also be tokenised the same way, token.IDENT, token.IDENT, token.ADD, token.IDENT, what it will do instead is that when the parser looks at the first token.IDENT (it figures out that this is a command) further to this all other operators (such as non command tokens (ADD, SUBTRACT etc) would just become the normal IDENT's as well. This makes it straight forward to make sure that operators like these `>>` which are command level operators don't get ignored by the parser.

Here is the problem with this approach:
- All command args are not valid idents. An argument like `123abc` will be lexed into `FLOAT(123)` and `IDENT(abc)`, though they are a single argument.
- Statements support only a subset of unicode characters, and putting other characters as a command argument will result in a `ILLEGAL` token. So something like `echo ൧` will emit `IDENT(echo)` and `ILLEGAL(൧)`.
- `echo a+b` and `echo a + b` will both be lexed into `echo`, `a`, `+`, `b`. Notice that in the first `a+b` is a single argument, as `+` is not a command operator. So instead of a single argument in the first example, we are wrongly lexing it into 3.

Reply:
> All command args are not valid idents. An argument like 123abc will be lexed into FLOAT(123) and IDENT(abc), though they are a single argument.

This first point is exactly why I didn't wanted to have two separate branches (what you call command and not command). Which is why, at the time I had recommended to leave all of these as IDENTS until we have more context in the parser and let the parser handle this since it will know that a group of tokens belong to the command branch of the decision or not a command branch.

> Statements support only a subset of unicode characters, and putting other characters as a command argument will result in a ILLEGAL token. So something like echo ൧ will emit IDENT(echo) and ILLEGAL(൧).

I don't understand what you mean here. So far, I have been under the impressions that statements support utf-8, for example, `let ൧ := 1` is supported. If this is not supported in the current implementation then I don't think it correctly handles utf-8 as mentioned in the PR description.

> echo a+b and echo a + b will both be lexed into echo, a, +, b. Notice that in the first a+b is a single argument, as + is not a command operator. So instead of a single argument in the first example, we are wrongly lexing it into 3.

I thought about this when I made the lexer previously and this was why I essentially went with the IDENTs only approach but again, this doesn't get solved even in the current implementation because this `echo a+b>>README` gets tokenised as IDENT(echo) and IDENT(a+b>>README). The way I wanted to implement it was, to have two lexers since we have two branches. One to work with commands and the other to work with language.

Also, if you add the test case 

```
86 echo a+b>>README
```


```
222		{token.SEMICOLON, "\n", 75, 17},
223		{token.STRING, "echo", 76, 1},
224		{token.STRING, "a+b", 76, 6},
225		{token.SHR, ">>", 76, 9},
226		{token.STRING, "README", 76, 11},
227		{token.SEMICOLON, "\n", 76, 18},
228		{token.EOF, "", 77, 1},
```

You will see that the current implementation is also not correct the way you've described it. Remember that `>>` is a command level token but that is not being token into consideration - precisely because it needs more context than just the last token which it doesn't have.

Reply:
The reason `echo hi>>a.txt` does not work because I thought it should not actually. I was under the impression that `echo a&&b` will emit `a&&b` but it turns out otherwise. It is a trivial fix, I will fix it so that it parses them correctly.

> I don't understand what you mean here. So far, I have been under the impressions that statements support utf-8, for example, let ൧ := 1 is supported. If this is not supported in the current implementation then I don't think it correctly handles utf-8 as mentioned in the PR description.

We support all the unicode characters with gc = L, which is what the `unicode.IsLetter` function says. `൧` is a number from another script(gc = N) and is currently not supported as an identifier.

Reply:
Should I resolve this @tjgurwara99?

Reply:
This can be marked resolved but I still think this needs more thought. The fundamental issue for this problem (to me) seems to be related to the command vs not command approach. There are other language where a `Program` (top level node in the `ast`) consists of more than one type of thing, but almost all of the elegant solutions consist of a keyword or token that distinguishes them from each other, take `func` or `type` or `=` as an example. These tokens or keywords make the program deterministic, we know that its one or the other but it can't be both. Maybe your solution would be nice, I still see that this would create problems but lets wait and see how this progresses in the parser.